File: capacity-to-ship-packages-within-d-days.py
class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        l = max(sum(weights) // days, max(weights))
        r = sum(weights)
        def isValid(sums : int) -> bool:
            splits = 0
            curr = 0
            for pre in weights:
                if curr + pre > sums:
                    splits += 1
                    curr = 0
                curr += pre
            return splits <= days - 1
        while l < r:
            m = (l + r) // 2
            if isValid(m):
                r = m
            else:
                l = m + 1
        return l

File: capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    bool isvalid(vector<int>& weights, int m, int days) {
        int count = 0, sum = 0;
        for(int i : weights) {
            if(sum + i > m) {
                count++;
                sum = 0;
            } 
            sum += i;
        }
        return count <= days - 1;
    }
    int shipWithinDays(vector<int>& weights, int days) {
        int sum = accumulate(weights.begin(), weights.end(), 0);
        int l = max(sum / days, *max_element(weights.begin(), weights.end())), r = sum;
        while(l < r) {
            int m = l + (r-l) / 2;
            if(!isvalid(weights, m, days)) {
                l = m + 1;
            } else {
                r = m;
            }
        }
        return l;
    }

};
